tests/components/influxdb/test_init.py

```diff
@@ -11,749 +11,1102 @@
+    """Test the setup with minimal configuration and defaults."""
+    config = {"influxdb": {}}
+    config["influxdb"].update(config_ext)
+
+    assert await async_setup_component(hass, influxdb.DOMAIN, config)
+    await hass.async_block_till_done()
+    assert hass.bus.listen.called
+    assert EVENT_STATE_CHANGED == hass.bus.listen.call_args_list[0][0][0]
+    assert get_write_api(mock_client).call_count == 1
+
+
+@pytest.mark.parametrize(
+    "mock_client, config_ext, get_write_api",
+    [
+        (influxdb.DEFAULT_API_VERSION, {"username": "user"}, _get_write_api_mock_v1),
+        (influxdb.DEFAULT_API_VERSION, {"token": "token"}, _get_write_api_mock_v1),
+        (
+            influxdb.API_VERSION_2,
+            {"api_version": influxdb.API_VERSION_2, "organization": "organization"},
+            _get_write_api_mock_v2,
+        ),
+        (
+            influxdb.API_VERSION_2,
+            {
+                "api_version": influxdb.API_VERSION_2,
+                "token": "token",
+                "organization": "organization",
                 "username": "user",
                 "password": "pass",
-                "exclude": {
-                    "entities": ["fake.blacklisted"],
-                    "domains": ["another_fake"],
-                },
-            }
+            },
+            _get_write_api_mock_v2,
+        ),
+    ],
+    indirect=["mock_client"],
+)
+async def test_invalid_config(hass, mock_client, config_ext, get_write_api):
+    """Test the setup with invalid config or config options specified for wrong version."""
+    config = {"influxdb": {}}
+    config["influxdb"].update(config_ext)
+
+    assert not await async_setup_component(hass, influxdb.DOMAIN, config)
+
+
+async def _setup(hass, mock_influx_client, config_ext, get_write_api):
+    """Prepare client for next test and return event handler method."""
+    config = {
+        "influxdb": {
+            "host": "host",
+            "exclude": {"entities": ["fake.blacklisted"], "domains": ["another_fake"]},
         }
-        config["influxdb"].update(kwargs)
-        assert setup_component(self.hass, influxdb.DOMAIN, config)
-        self.handler_method = self.hass.bus.listen.call_args_list[0][0][1]
-        mock_client.return_value.write_points.reset_mock()
-
-    def test_event_listener(self, mock_client):
-        """Test the event listener."""
-        self._setup(mock_client)
-
-        # map of HA State to valid influxdb [state, value] fields
-        valid = {
-            "1": [None, 1],
-            "1.0": [None, 1.0],
-            STATE_ON: [STATE_ON, 1],
-            STATE_OFF: [STATE_OFF, 0],
-            STATE_STANDBY: [STATE_STANDBY, None],
-            "foo": ["foo", None],
+    }
+    config["influxdb"].update(config_ext)
+    assert await async_setup_component(hass, influxdb.DOMAIN, config)
+    # A call is made to the write API during setup to test the connection.
```

Comment:
We should block till done here too with `hass.async_block_till_done`.

Reply:
Yep, done. Did a quick scan and every `async_setup_component` call is now followed by a `hass.async_block_till_done`. Well except the one test that tests failure scenarios for setting up the component.